File: compiler/src/core/macrocosm.py
from contextlib import contextmanager
from typing import Any, Sequence
from io import StringIO
from macros.noscope_macro import Noscope_macro_provider
from macros.collection_macros import List_macro_provider, Dict_macro_provider
from macros.if_macro import If_macro_provider
from macros.literal_value_macros import Number_macro_provider, String_macro_provider
from macros.while_macro import While_macro_provider
from macros.for_macro import For_macro_provider
from macros.local_macro import Local_macro_provider
from macros.fn_macro import Fn_macro_provider
from macros.exists_macro import Exists_macro_provider
from macros.then_macro import Then_macro_provider, Pipeline_macro_provider
from macros.multi_provider import Multi_provider
from macros.call_macro import Call_macro_provider
from macros.utility_macros import Noop_macro_provider
from macros.solution_macro import Solution_macro_provider
from macros.error_macros import Must_compile_error_macro_provider
from macros.comment_macros import Comment_macro_provider, COMMENT_MACROS
from macros.return_macro import Return_macro_provider
from macros.scope_macro import Scope_macro_provider, SCOPE_MACRO
from macros.type_macro import Type_macro_provider # Added import
from macros.try_catch_macro import Try_macro_provider, Catch_macro_provider, Finally_macro_provider, Throw_macro_provider
from macros.bind_macro import Bind_macro_provider
from macros.obtain_param_value_macro import Obtain_param_value_macro_provider
from core.node import Node, Position, Macro, Args
from utils.strutil import IndentedStringIO, Joiner
from pipeline.steps import MacroProcessingStep
from core.exceptions import MacroAssertFailed
from pipeline.js_conversion import to_valid_js_ident
from core.macro_registry import Macro_emission_provider, MacroContext, Macro_provider, MacroRegistry
from pipeline.steps import PreprocessingStep
from linking.code_block_linking import CodeBlockLinkingStep  
from pipeline.steps import TypeCheckingStep
from pipeline.steps import TypeRegistrationStep
from pipeline.steps.type_registration import TypeDetailRegistrationStep
from pipeline.steps import MustCompileErrorVerificationStep
from pipeline.steps import JavaScriptEmissionStep
from utils.logger import default_logger

class Macrocosm:
    def __init__(self, emission_registry: MacroRegistry, typecheck_registry: MacroRegistry, code_linking_registry: MacroRegistry, preprocess_registry: MacroRegistry, type_registration_registry: MacroRegistry, type_detail_registration_registry: MacroRegistry):
        self.nodes: list[Node] = []
        # TODO. incremental is good enough for now, but we'll have to stabilize it.
        #  the last thing you would want is the entire output changing because you added a statement. that's a lot of
        #  unnecessary diff. best way to solve this that i see is to make this block-scoped,
        #  so that each block gets its own incremental. a bit harder, though.
        self.incremental_id = 0
        self.compile_errors: list[dict[str, Any]] = []
        self._js_output: str = ""
        
        # Metadata is now stored directly on Node objects via setattr

        # Dynamic call conventions for user-defined types
        self._dynamic_conventions: dict[str, list[Any]] = {}

        self.root_node: Node | None = None
        
        # Initialize the processing pipeline
        self.processing_steps: list[MacroProcessingStep] = [
            CodeBlockLinkingStep(code_linking_registry), 
            PreprocessingStep(preprocess_registry),
            TypeRegistrationStep(type_registration_registry),     # Pass 1: Register basic type shells
            TypeDetailRegistrationStep(type_detail_registration_registry), # Pass 2: Fill in field details
            TypeCheckingStep(typecheck_registry),
            JavaScriptEmissionStep(emission_registry),
            MustCompileErrorVerificationStep()
        ]

        self.registries: dict[str, MacroRegistry] = {}

    def get_new_ident(self, name: str | None):
        ident = f"_{hex(self.incremental_id)}"
        if name:
            ident += f"_{to_valid_js_ident(name)}"
        self.incremental_id += 1
        return ident

    def get_metadata(self, node: Node, metadata_type: type):
        """Get metadata for a node, auto-computing Macro and Args if missing"""
        default_logger.log("metadata", f"get metadata {str(metadata_type)} for {id(node)} {node.content}")
        
        # Use getattr to retrieve metadata directly from the node object
        attr_name = f"_metadata_{metadata_type.__name__}"
        
        # Auto-compute Macro and Args if not present
        if metadata_type in [Macro, Args] and not hasattr(node, attr_name):
            self._ensure_macro_args_computed(node)
        
        if hasattr(node, attr_name):
            return getattr(node, attr_name)
        
        # Check if there's a default factory from the old TypeMap system
        from utils.utils import TypeMap
        if metadata_type in TypeMap._default_factories:
            value = TypeMap._default_factories[metadata_type]()
            self.set_metadata(node, metadata_type, value)
            return value
        
        raise KeyError(f"No metadata of type {metadata_type} for node")

    def maybe_metadata(self, node: Node, metadata_type: type):
        """Get metadata for a node if it exists, return None otherwise"""
        try:
            return self.get_metadata(node, metadata_type)
        except KeyError:
            return None

    def set_metadata(self, node: Node, metadata_type: type, value: Any):
        """Set metadata for a node"""
        # Use setattr to store metadata directly on the node object
        attr_name = f"_metadata_{metadata_type.__name__}"
        setattr(node, attr_name, value)
        default_logger.log("metadata", f"set metadata {str(metadata_type)} {str(value)} for {id(node)} {node.content}")

    def invalidate_metadata(self, node: Node):
        """Invalidate metadata for a node and all its descendants when tree changes"""
        # Remove all metadata attributes from the node
        attrs_to_remove = [attr for attr in dir(node) if attr.startswith('_metadata_')]
        for attr in attrs_to_remove:
            delattr(node, attr)
        
        # Recursively invalidate children
        for child in node.children:
            self.invalidate_metadata(child)

    def _ensure_macro_args_computed(self, node: Node):
        """Ensure Macro and Args metadata is computed for a node"""
        from core.node import Macro, Args
        from utils.strutil import cut
        
        macro, args = cut(node.content, " ")
        self.set_metadata(node, Macro, macro)
        self.set_metadata(node, Args, args)

    def register(self, node: Node):
        self.nodes.append(node)

    def assert_(self, must_be_true: bool, node: Node, message: str, error_type: str = None, extra_fields: dict[str, Any] = None):
        if not must_be_true:
            from utils.error_types import ErrorType
            if error_type is None:
                error_type = ErrorType.ASSERTION_FAILED
            self.compile_error(node, f"failed to assert: {message}", error_type, extra_fields)
            raise MacroAssertFailed(message)

    def compile_error(self, node: Node, error: str, error_type: str, extra_fields: dict[str, Any] = None):
        """Add a compile error with explicit error type."""
        pos = node.pos or Position(0, 0)
        entry: dict[str, Any] = { # TODO dataclass
            "recoverable": False, # TODO
            "line": pos.line,
            "char": pos.char,
            "content": node.content,
            "error": error,
            "error_type": error_type
        }
        if extra_fields:
            entry.update(extra_fields)
        self.compile_errors.append(entry)

    def compile(self):
        # Discover macros first
        with default_logger.indent("compile", "discovering macros"):
            for node in self.nodes:
                self.__discover_macros(node)
            
        solution_node = self.make_node("67lang:solution", Position(0, 0), self.nodes or [])
        self.root_node = solution_node
            
        # Execute the processing pipeline
        for step in self.processing_steps:
            step_name = step.__class__.__name__
            with default_logger.indent("compile", f"processing step: {step_name}"):
                ctx = MacroContext(
                    statement_out=StringIO(),  # dummy for non-emission steps
                    expression_out=StringIO(),
                    node=solution_node,
                    compiler=self,
                    current_step=step,
                )
                step.process_node(ctx)
        
        if len(self.compile_errors) != 0:
            return "" # TODO - raise an error instead ?
        
        return self._js_output

    def __discover_macros(self, node: Node):
        # TODO lstring macros should perhaps get special handling here...
        self._ensure_macro_args_computed(node)
        for child in node.children:
            self.__discover_macros(child)

    def make_node(self, content: str, pos: Position, children: None | list[Node]) -> Node:
        n = Node(content, pos, children)
        self.__discover_macros(n)
        return n

    def add_dynamic_convention(self, name: str, convention: Any):
        if name not in self._dynamic_conventions:
            self._dynamic_conventions[name] = []
        self._dynamic_conventions[name].append(convention)

    # TODO - probably time to nuke this one...
    def compile_fn_call(self, ctx: MacroContext, call: str, nodes: Sequence[Node], ident:bool=True):
        from dataclasses import replace
        args: list[str] = []
        for child in nodes:
            expression_out = IndentedStringIO()
            child_ctx = replace(ctx, node=child, expression_out=expression_out)
            child_ctx.current_step.process_node(child_ctx)
            expression_out = expression_out.getvalue()
            if expression_out:
                args.append(expression_out)
            
        ident_value = ""
        if ident:
            ident_value = ctx.compiler.get_new_ident(call)
            ctx.statement_out.write(f"const {ident_value} = ")
        ctx.statement_out.write(f"{call}")
        joiner = Joiner(ctx.statement_out, ", ")
        for i in args:
            with joiner:
                ctx.statement_out.write(i)
        ctx.statement_out.write(")\n")
        if ident:
            ctx.expression_out.write(ident_value)

    @property
    def safely(self):
        @contextmanager
        def _safely():
            try:
                yield
            except MacroAssertFailed:
                pass
        return _safely()

def create_macrocosm() -> Macrocosm:
    # creates it with all the necessary macros registered
    
    def has_arguments(ctx: MacroContext) -> bool:
        """Matcher function: returns True if the macro has arguments"""
        from core.node import Args
        args = ctx.compiler.get_metadata(ctx.node, Args)
        return bool(args.strip())
    
    macro_providers: dict[str, Macro_provider] = {
        "while": While_macro_provider(),
        "for": For_macro_provider(),
        "int": Number_macro_provider(int),
        "float": Number_macro_provider(float),
        "string": String_macro_provider("string"),
        "regex": String_macro_provider("regex"),
        "if": If_macro_provider(),
        "list": List_macro_provider(),
        "dict": Dict_macro_provider(),
        "local": Local_macro_provider(),
        "fn": Fn_macro_provider(),
        "bind": Bind_macro_provider(),
        
        
        "67lang:call": Call_macro_provider(),
        "exists": Exists_macro_provider(),
        "noop": Noop_macro_provider(),
        "type": Type_macro_provider(),
        "67lang:assume_local_exists": Noop_macro_provider(),
        "67lang:obtain_param_value": Obtain_param_value_macro_provider(),
        "67lang:last_then": Noop_macro_provider(),
        "67lang:solution": Solution_macro_provider(),
        "must_compile_error": Must_compile_error_macro_provider(),
        "then": Multi_provider([
            (has_arguments, Pipeline_macro_provider()),
            (None, Scope_macro_provider())
        ]),
        "do": Multi_provider([
            (has_arguments, Pipeline_macro_provider()),
            (None, Scope_macro_provider())
        ]),
        "get": Pipeline_macro_provider(),
        "noscope": Noscope_macro_provider(),
        "return": Return_macro_provider(),
        "try": Try_macro_provider(),
        "catch": Catch_macro_provider(),
        "finally": Finally_macro_provider(),
        "throw": Throw_macro_provider(),
        
        # List operation macros (handled by List_macro_provider)
        "append": Noop_macro_provider(),
        "prepend": Noop_macro_provider(),
        "insert_after_index": Noop_macro_provider(),
        
        # Dict entry macro (handled by Dict_macro_provider)
        "entry": Noop_macro_provider(),
    }

    for macro in COMMENT_MACROS:
        macro_providers[macro] = Comment_macro_provider()

    for macro in SCOPE_MACRO:
        macro_providers[macro] = Scope_macro_provider()

    literally = {
        "true": ("true", "bool"),
        "false": ("false", "bool"), 
        "break": ("break", None),
        "continue": ("continue", None)
    }

    # TODO, this should be... elsewhere.
    class Literal_macro_provider(Macro_emission_provider):
        def __init__(self, value: str, type_name: str = None):
            self.value = value
            self.type_name = type_name
        def emission(self, ctx: MacroContext):
            ctx.expression_out.write(self.value)
        def typecheck(self, ctx: MacroContext):
            if self.type_name:
                return self.type_name
    
    for k, (value, type_name) in literally.items():
        macro_providers[k] = Literal_macro_provider(value, type_name)

    for name, provider in macro_providers.items():
        default_logger.registry(f"registering macro '{name}' -> {provider.__class__.__name__}")

    registries = {}
    def create_registry(name: str):
        registry = MacroRegistry()
        registries[name] = registry
        return registry
    
    
    # how FUCKING ironic to have to write such code in a language
    # that doesn't offer macros only to create a new language
    # that would trivially implement this with zero repetition?

    preprocess = create_registry("preprocess")
    typecheck = create_registry("typecheck")
    emission = create_registry("emission")
    type_registration = create_registry("type_registration")
    type_detail_registration = create_registry("type_detail_registration")
    
    # this is the new way of doing things
    # we still need to bridge this to the old way, for now
    # TODO - remove the old way
    
    code_linking_registry = create_registry("code_linking")

    for macro, provider in macro_providers.items():
        preprocess.add_fn(getattr(provider, "preprocess", None), macro)
        typecheck.add_fn(getattr(provider, "typecheck", None), macro)
        emission.add_fn(getattr(provider, "emission", None), macro)
        type_registration.add_fn(getattr(provider, "register_type", None), macro)
        type_detail_registration.add_fn(getattr(provider, "register_type_details", None), macro)
        code_linking_registry.add_fn(getattr(provider, "code_linking", None), macro)  
    
    rv = Macrocosm(emission, typecheck, code_linking_registry, preprocess, type_registration, type_detail_registration)
    rv.registries.update(registries)
    return rv
    

File: compiler/src_ts/core/macrocosm.ts
// core/macrocosm.ts

import { cut } from "../utils/strutil.ts";
import { default_logger } from "../utils/logger.ts";
import {
  Node,
  Position,
  Macro,
  Args,
  Params,
  Indexers,
  Callers
} from "./node.ts";
import { Constructor, not_null, TypeMap } from "../utils/utils.ts";
import { to_valid_js_ident } from "../pipeline/js_conversion.ts";
import { Macro_error_tracker } from "../core/exceptions.ts";
import { Macro_context, Macro_registry, Register_macro_providers, Macro_provider, REGISTER_MACRO_PROVIDERS, Ctx_proc } from "./macro_registry.ts";

import { Noscope_macro_provider } from "../macros/noscope_macro.ts";
import {
  List_macro_provider,
  Dict_macro_provider,
} from "../macros/collection_macros.ts";
import { If_macro_provider } from "../macros/if_macro.ts";
import {
  Number_macro_provider,
  String_macro_provider,
} from "../macros/literal_value_macros.ts";
import { While_macro_provider } from "../macros/while_macro.ts";
import { For_macro_provider } from "../macros/for_macro.ts";
import { Local_macro_provider } from "../macros/local_macro.ts";
import { Fn_macro_provider } from "../macros/fn_macro.ts";
import { Call_macro_provider } from "../macros/call_macro.ts";
import {
  Comment_macro_provider,
} from "../macros/comment_macros.ts";
import { Return_macro_provider } from "../macros/return_macro.ts";
import {
  Scope_macro_provider,
} from "../macros/scope_macro.ts";
import { Type_macro_provider } from "../macros/type_macro.ts";
import {
  Try_macro_provider,
  Throw_macro_provider,
} from "../macros/try_catch_macro.ts";
import { Bind_macro_provider } from "../macros/bind_macro.ts";
import {
  Must_compile_error_macro_provider,
} from "../macros/error_macros.ts";

import {
  Expression_return_type,
  Type_check_result,
  Type_engine,
} from "../compiler_types/proper_types.ts";
import { JSON_value } from "./meta_value.ts";
import { Preprocessing_context } from "../pipeline/steps/processing.ts";
import { Must_compile_error_context } from "../pipeline/steps/must_compile_error_step.ts";
import { Solution_macro_provider } from "../macros/solution_macro.ts";
import { Multi_provider } from "../macros/multi_provider.ts";
import { _67lang_GET_LAST_PIPELINE_RESULT, _67lang_PIPELINE_RESULT, Get_last_pipeline_result_macro_provider, Pipeline_macro_provider, Pipeline_result_macro_provider } from "../macros/then_macro.ts";
import { Noop_macro_provider } from "../macros/utility_macros.ts";
import { Type_checking_context, Type_registration_context } from "../pipeline/steps/typechecking.ts";
import { Emission_macro_context } from "../pipeline/steps/emission.ts";

// --- metadata store (TS replacement for def_metadata macro) -------------------

export type MetaCtor<T> = { new (...args: any[]): T; name: string };

export class Macrocosm {
  readonly nodes: Node[] = [];
  compile_errors: Array<Record<string, JSON_value>> = [];

  private incremental_id = 0;
  js_output = "";
  root_node: Node | null = null;

  readonly processing_steps: ((node: Node) => Macro_context)[];

  readonly error_tracker: Macro_error_tracker = new Macro_error_tracker();

  // metadata: Node -> (type-name -> value)
  private metadata: Map<Node, Map<string, unknown>> = new Map();

  constructor(
    emission_registry: Macro_registry<Emission_macro_context>,
    typecheck_registry: Macro_registry<Type_checking_context>,
    preprocess_registry: Macro_registry<Preprocessing_context>,
    type_registration_registry: Macro_registry<Type_registration_context>,
  ) {
    const compiler = this;
    const type_engine = new Type_engine();
    this.processing_steps = [
      (node: Node) => new Preprocessing_context({ compiler, registry: preprocess_registry, node }),
      (node: Node) => new Type_registration_context({ compiler, registry: type_registration_registry, node, type_engine }),
      (node: Node) => new Type_checking_context({ compiler, registry: typecheck_registry, node, type_engine }),
      (node: Node) => new Emission_macro_context({ compiler, registry: emission_registry, statement_out: [], expression_out: [], node, type_engine }),
      (node: Node) => new Must_compile_error_context({ compiler, registry: new Macro_registry<Must_compile_error_context>(), node, expectations: [] }), // TODO: share registry
    ];
  }

  get_new_ident(name: string | null): string {
    let ident = `_0x${this.incremental_id.toString(16)}`;
    if (name && name.length > 0) {
      ident += `_${to_valid_js_ident(name)}`;
    }
    this.incremental_id += 1;
    return ident;
  }

  // --- metadata APIs (typed) --------------------------------------------------

  private metaKey<T>(ctor: MetaCtor<T>): string {
    return ctor.name;
  }

  set_metadata<T>(node: Node, t: MetaCtor<T>, value: T): void {
    const key = this.metaKey(t);
    let m = this.metadata.get(node);
    if (!m) {
      m = new Map();
      this.metadata.set(node, m);
    }
    m.set(key, value);
    default_logger.log(
      "metadata",
      `set metadata ${key} ${String(value)} for ${node} ${node.content}`,
    );
  }

  maybe_metadata<T>(node: Node, t: MetaCtor<T>): T | null {
    const key = this.metaKey(t);
    const m = this.metadata.get(node);
    if (!m) {
      return null;
    }
    const v = m.get(key);
    return (v as T | undefined) ?? null;
  }

  compute_metadata<T>(node: Node, t: MetaCtor<T>, factory: () => T): T {
    const existing = this.maybe_metadata(node, t);
    if (existing !== null) {
      return existing;
    }
    const v = factory();
    this.set_metadata(node, t, v);
    return v;
  }

  get_metadata<T>(node: Node, t: MetaCtor<T>): T {
    // special-case Macro / Args to lazily compute from node.content
    if (t === (Macro as unknown)) {
      this.ensure_macro_args_computed(node);
    }
    if (t === (Args as unknown)) {
      this.ensure_macro_args_computed(node);
    }

    const key = this.metaKey(t);
    const m = this.metadata.get(node);
    if (m && m.has(key)) {
      return m.get(key) as T;
    }

    const factory = TypeMap.getFactory(t);
    if (factory) {
      const v = factory();
      let mm = this.metadata.get(node);
      if (!mm) {
        mm = new Map();
        this.metadata.set(node, mm);
      }
      mm.set(key, v);
      return v;
    }

    throw new Error(`No metadata of type ${key} for node`);
  }

  invalidate_metadata(node: Node): void {
    this.metadata.delete(node);
    for (const ch of node.children) {
      this.invalidate_metadata(ch);
    }
  }

  // --- macro arg splitting ----------------------------------------------------

  private ensure_macro_args_computed(node: Node): void {
    const hasMacro = this.maybe_metadata(node, Macro) !== null;
    const hasArgs = this.maybe_metadata(node, Args) !== null;
    if (hasMacro && hasArgs) {
      return;
    }

    const [macro_name, args] = cut(node.content, " ");
    this.set_metadata(node, Macro, new Macro(macro_name));
    this.set_metadata(node, Args, new Args(args));
  }

  // --- public API -------------------------------------------------------------

  register(node: Node): void {
    this.nodes.push(node);
  }

  private compile_error(
    node: Node,
    error: string,
    error_type: string,
    extra_fields: Record<string, JSON_value> | null = null,
  ): void {
    const pos = node.pos ?? new Position(0, 0);
    const entry: Record<string, JSON_value> = {
      recoverable: false as unknown as JSON_value,
      line: pos.line as unknown as JSON_value,
      char: pos.char as unknown as JSON_value,
      content: node.content as unknown as JSON_value,
      error: error as unknown as JSON_value,
      error_type: error_type as unknown as JSON_value,
    };

    if (extra_fields) {
      for (const [k, v] of Object.entries(extra_fields)) {
        entry[k] = v;
      }
    }

    this.compile_errors.push(entry);
  }

  compile(): string {
    default_logger.indent("compile", "discovering macros", () => {
      for (const n of this.nodes) {
        this.discover_macros(n);
      }
    });

    const solution_node = this.make_node(
      "67lang:solution",
      new Position(0, 0),
      [...this.nodes],
    );
    this.root_node = solution_node;

    for (const step of this.processing_steps) {
      // TODO
      // const stepName = (step as any).constructor.name;
      // default_logger.indent(
      //   "compile",
      //   `processing step: ${stepName}`,
        // () => {
        const ctx = step(solution_node)
        this.error_tracker.safely(ctx, () => {
          ctx.apply();
        });
        // },
      // );
    }

    for (const err of this.error_tracker.all) {
      this.compile_error(
        err.node,
        err.message,
        err.type,
        err.extras ?? null,
      );
    }

    if (this.compile_errors.length !== 0) {
      return "";
    }

    return this.js_output;
  }

  private discover_macros(node: Node): void {
    this.ensure_macro_args_computed(node);
    for (const ch of node.children) {
      this.discover_macros(ch);
    }
  }

  make_node(
    content: string,
    pos: Position | null,
    children: Node[] | null,
  ): Node {
    const p = pos ?? new Position(0, 0);
    const n = new Node(content, p, children ?? []);
    this.discover_macros(n);
    return n;
  }

}

// Literal_macro_provider

export class Literal_macro_provider implements Macro_provider {
  constructor(
    private macro_name: string, 
    private value: string, 
    private t_name: string | null
  ) {
    // ...
  }

  [REGISTER_MACRO_PROVIDERS](via: Register_macro_providers): void {
    via(Emission_macro_context, this.macro_name, this.emission.bind(this));
    via(Type_checking_context, this.macro_name, this.typecheck.bind(this));
  }

  emission(ctx: Emission_macro_context): void {
    if (this.t_name === null) {
      ctx.statement_out.push(() => this.value);
    } else {
      ctx.expression_out.push(() => this.value);
    }
  }

  typecheck(ctx: Type_checking_context): Type_check_result {
    if (this.t_name) {
      return new Expression_return_type(ctx.type_engine.get_type(this.t_name));
    } else {
      // TODO a stupid h*ck
      return null;
    }    
  }
}

// --- macrocosm factory --------------------------------------------------------

export function create_macrocosm(): Macrocosm {
  const has_arguments = (ctx: Macro_context): boolean => {
    const args = ctx.compiler.get_metadata(ctx.node, Args).toString();
    return args.trim().length > 0;
  };

  const providers = [
    new Noop_macro_provider(),
    new Type_macro_provider(),
    new While_macro_provider(),
    new For_macro_provider(),
    new Number_macro_provider("int"),
    new Number_macro_provider("float"),
    new String_macro_provider("string"),
    new String_macro_provider("regex"),
    new If_macro_provider(),
    new List_macro_provider(),
    new Dict_macro_provider(),
    new Local_macro_provider(),
    new Fn_macro_provider(),
    new Bind_macro_provider(),
    new Call_macro_provider(),
    new Pipeline_result_macro_provider(),
    new Get_last_pipeline_result_macro_provider(),
    new Solution_macro_provider(),
    new Must_compile_error_macro_provider(),
    new Comment_macro_provider(),
    new Multi_provider(
      "do",
      [
        [has_arguments, new Pipeline_macro_provider()],
        [null, new Scope_macro_provider()]
      ],
    ),
    new Multi_provider(
      "then",
      [
        [has_arguments, new Pipeline_macro_provider()],
        [null, new Scope_macro_provider()]
      ],
    ),
    new Multi_provider("get", [[null, new Pipeline_macro_provider()]]),
    new Multi_provider("set", [[null, new Pipeline_macro_provider()]]),
    new Multi_provider("67lang:file", [[null, new Scope_macro_provider()]]),
    new Multi_provider("finally", [[null, new Scope_macro_provider()]]),    
    new Noscope_macro_provider(),
    new Return_macro_provider(),
    new Try_macro_provider(),
    new Throw_macro_provider()
  ]
  
  const literalTable: Record<string, [string, string | null]> = {
    true: ["true", "bool"],
    false: ["false", "bool"],
    break: ["break;", null],
    continue: ["continue;", null],
  };

  for (const [macro_name, [value, type]] of Object.entries(literalTable)) {
    providers.push(new Literal_macro_provider(macro_name, value, type));
  }

  const registries: Map<Constructor<Macro_context>, Macro_registry<Macro_context<unknown>>> = new Map();
  const create_registry = <T extends Macro_context<unknown>>(_name: string, type: Constructor<T>): Macro_registry<T> => {
    const r = new Macro_registry<T>();
    registries.set(type, r as unknown as Macro_registry<Macro_context>);
    return r;
  };

  const preprocess = create_registry<Preprocessing_context>("preprocess", Preprocessing_context);
  const typecheck = create_registry<Type_checking_context>("typecheck", Type_checking_context);
  const emission = create_registry<Emission_macro_context>("emission", Emission_macro_context);
  const type_registration = create_registry<Type_registration_context>("type_registration", Type_registration_context);
  for (const p of providers) {
    p[REGISTER_MACRO_PROVIDERS]((
      ctxType,
      macro_name,
      proc,
    ) => {
      const registry = not_null(registries.get(ctxType));
      registry.add_fn(proc as Ctx_proc<Macro_context>, macro_name);
    });
  }

  const rv = new Macrocosm(
    emission,
    typecheck,
    preprocess,
    type_registration,
  );
  return rv;
}

// default factories for metadata types that want auto-creation

TypeMap.register(Callers, () => new Callers());
TypeMap.register(Indexers, () => new Indexers());
TypeMap.register(Params, () => new Params());
